### PR TITLE
refactor: Reuse block hash in AddToBlockIndex

### PR DESCRIPTION
**Pull request by promag**
Removes block hash computation from `AddToBlockIndex`. Apart from `LoadGenesisBlock`, the hash is already computed in `AcceptBlockHeader`.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Tend to NACK. This makes the code harder to read

**Comment by promag** [comment]
I don't think it makes the code harder to read honestly. This saves a bit of CPU for each new `CBlockIndex` since `CBlockHeader::GetHash()` is not cached.

**Comment by jb55** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
Instead of passing the block to `AddToBlockIndex`, you'd have to pass the hash and the block. Unintuitive and easy to accidentally pass the wrong hash. If this is a performance optimization, maybe the hash should be cached? And it needs a benchmark either way.

**Comment by promag** [comment]
@MarcoFalke fair enough, I'll see if such change is worth it.

**Comment by promag** [comment]
@MarcoFalke let me know what you think.

**Comment by jonasschnelli** [comment]
We are talking only about the block header hash? @promag: can you "quantify" the speedup?

File: src/validation.cpp [hunk 1]
```diff
@@ -60,6 +60,17 @@
 #define MICRO 0.000001
 #define MILLI 0.001
 
+struct BlockHeaderHashed
+{
+    const CBlockHeader& header;
+    const uint256 hash;
+    BlockHeaderHashed(const CBlockHeader& header)
```
Comment (practicalswift) [inline src/validation.cpp]: `explicit`? :-)

**Comment by promag** [comment]
Profiling was done in debug mode, going to close and reopen if relevant in release mode.

**Comment by jamesob** [comment]
Also tend to NACK unless we can show significant performance impact. Introducing an additional identity (`BlockHeaderHashed`) makes the code slightly more complicated, especially when introduced without any documentation (when should this be used and why?).

**Comment by promag** [comment]
Before reopening I'd like to share some profile results. The following is the result of running in release mode, with an empty `-datadir` before and after. It shows the total time spent on `ProcessNewBlockHeaders` (currently 574000 headers). As you can see reusing the hash reduces the total time from 2.44s to 1.67s.

Before:
Screenshot 2019-05-22 at 16 26 12
Screenshot 2019-05-22 at 16 20 11

After:
Screenshot 2019-05-22 at 16 25 57

Screenshot 2019-05-22 at 16 20 29

**Pull request closed without merging**